posts/scala-for-the-arrogant-part-2.md

```diff
@@ -0,0 +1,427 @@
+
+# Devising A Solution
+First we consider the elements of our solution (before starting, you might want to try and devise this yourself as it's a fun problem to tackle and you most likely do not have an armed explosive sitting nearby if you make a mistake. [The starting commit is here](https://github.com/randeepbydesign/diehardJugSolver/commit/8fe691efaaa7f4135f81bf0b3b0765f279439148)). The assumptions are:
+
+- A virtually unlimited supply of water
+- Can use two or more jugs of known size
+- The goal amount must be less than the capacity of the largest jug (a solution would be impossible otherwise)
+- No equipment for measuring until the end, meaning we have to fill a jug up all the way or not at all (no filling up a jug halfway)
+
+With these assumptions laid out, let's begin!
+
+## Define our problem...
+We start by thinking about the landscape of our problem and what associated objects we can create to implement our solution. The first thing that comes to mind are the **jugs**. We could model an object with a capacity and an amount:
+
+```java
+public class Jug {
+    final Integer capacity;
+    Integer currentAmount;
+    //Define getters and setters
+}
+```
+However, the capacity is fixed and the amount fluctuates. A better approach will be to define our Jugs as a list of Integers, and a separate list of Integers will specify how full they are at a given moment. We'll refer to that as the state (so if we start out with two empty jugs, we'd have a list [0, 0]). We then adjust the current amount based on the actions we take. Speaking of **Actions**, we should consider what movements we can take to solve the puzzle.
+
+In this case, our game is simple enough that we can only take three actions: 
+
+1. Fill up a jug to its full capacity
+2. Empty a jug so that its capacity is 0
+3. Pour the contents of one jug into another
+
+A simple enumeration defines these movements:
+
+```java
+public enum MoveType {
+    EMPTY,
+    FILL, 
+    POUR
+}
+```
+
+The **move** itself is modeled as an `Interface` with a single method `doMove` that takes some state, a pointer to the jug index or indexes that are relevant to the action, and returns the result of taking the action on the jug(s). Because `EMPTY` and `FILL` are single-jug operations they only receive one index for processing while `POUR` requires two: the jug we are pouring from and the jug we are pouring to. We pass our JUG capacities to this method as well so that we know the appropriate specifics for our FILL and POUR operations:
+
+`Integer[] doMove(final Integer[] JUGS, Integer[] state, Integer[] jugIndex);`
+
+For example, the implementation of the `FILL` action:
+```java
+Integer[] doMove(final Integer[] JUGS, Integer[] state, Integer[] jugIndex) {
```

Comment:
I don't know about that... I use it to designate constant variables as well. In this case there's nothing to prevent someone from modifying the elements of an array in Java so, at the very least, all Upper case would provide a visual indicator that this is some fixed data.

Reply:
See https://www.oracle.com/technetwork/java/javase/overview/codeconventions-135099.html

Reply:
I should add that "constant" doesn't mean "immutable". `final` in your code already expresses `constant`.

Reply:
I'm fine with leaving this one as is